### PR TITLE
quantum leap of CI technology that syncs with Google Colab and Python 3.11

### LINKED ISSUE #188: old matplotlib version breaks subplots
this code
```
fig, axes, _ = get_gridspec_subplots(2,1,wspace=0.3,hspace=0.3)
start_idx = 1 if exclude_first_trade is True else 0
first_trade_that_is_on_last_day = min(trades.index[trades.day == max(trades.day)])
end_idx = first_trade_that_is_on_last_day - 1 if exclude_last_day is True else len(trades)
axis = trades.iloc[start_idx:end_idx].plot(x="index", y=['share_reserves','bond_reserves','lp_reserves'], ax=axes[0])
axis = trades.iloc[start_idx:end_idx].plot(x="index", y=['share_reserves','bond_reserves','lp_reserves'], ax=axes[1])
```

errors as follows:
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[39], line 6
      4 first_trade_that_is_on_last_day = min(trades.index[trades.day == max(trades.day)])
      5 end_idx = first_trade_that_is_on_last_day - 1 if exclude_last_day is True else len(trades)
----> 6 axis = trades.iloc[start_idx:end_idx].plot(x="index", y=['share_reserves','bond_reserves','lp_reserves'], ax=axes[0])
      7 axis = trades.iloc[start_idx:end_idx].plot(x="index", y=['share_reserves','bond_reserves','lp_reserves'], ax=axes[1])
      8 # axis.get_legend().remove()

File ~/.pyenv/versions/3.8.16/envs/elf-env/lib/python3.8/site-packages/pandas/plotting/_core.py:1000, in PlotAccessor.__call__(self, *args, **kwargs)
    997             label_name = label_kw or data.columns
    998             data.columns = label_name
-> 1000 return plot_backend.plot(data, kind=kind, **kwargs)

File ~/.pyenv/versions/3.8.16/envs/elf-env/lib/python3.8/site-packages/pandas/plotting/_matplotlib/__init__.py:71, in plot(data, kind, **kwargs)
     69         kwargs["ax"] = getattr(ax, "left_ax", ax)
     70 plot_obj = PLOT_CLASSES[kind](data, **kwargs)
---> 71 plot_obj.generate()
     72 plot_obj.draw()
     73 return plot_obj.result

File ~/.pyenv/versions/3.8.16/envs/elf-env/lib/python3.8/site-packages/pandas/plotting/_matplotlib/core.py:455, in MPLPlot.generate(self)
    453 self._add_table()
    454 self._make_legend()
--> 455 self._adorn_subplots()
    457 for ax in self.axes:
    458     self._post_plot_logic_common(ax, self.data)

File ~/.pyenv/versions/3.8.16/envs/elf-env/lib/python3.8/site-packages/pandas/plotting/_matplotlib/core.py:679, in MPLPlot._adorn_subplots(self)
    677     all_axes = self._get_subplots()
    678     nrows, ncols = self._get_axes_layout()
--> 679     handle_shared_axes(
    680         axarr=all_axes,
    681         nplots=len(all_axes),
    682         naxes=nrows * ncols,
    683         nrows=nrows,
    684         ncols=ncols,
    685         sharex=self.sharex,
    686         sharey=self.sharey,
    687     )
    689 for ax in self.axes:
    690     ax = getattr(ax, "right_ax", ax)

File ~/.pyenv/versions/3.8.16/envs/elf-env/lib/python3.8/site-packages/pandas/plotting/_matplotlib/tools.py:410, in handle_shared_axes(axarr, nplots, naxes, nrows, ncols, sharex, sharey)
    408 layout = np.zeros((nrows + 1, ncols + 1), dtype=np.bool_)
    409 for ax in axarr:
--> 410     layout[row_num(ax), col_num(ax)] = ax.get_visible()
    412 for ax in axarr:
    413     # only the last row of subplots should get x labels -> all
    414     # other off layout handles the case that the subplot is
    415     # the last in the column, because below is no subplot/gap.
    416     if not layout[row_num(ax) + 1, col_num(ax)]:

File ~/.pyenv/versions/3.8.16/envs/elf-env/lib/python3.8/site-packages/pandas/plotting/_matplotlib/tools.py:396, in handle_shared_axes.<locals>.<lambda>(x)
    386 def handle_shared_axes(
    387     axarr: Iterable[Axes],
    388     nplots: int,
   (...)
    393     sharey: bool,
    394 ):
    395     if nplots > 1:
--> 396         row_num = lambda x: x.get_subplotspec().rowspan.start
    397         col_num = lambda x: x.get_subplotspec().colspan.start
    399         if compat.mpl_ge_3_4_0():

AttributeError: 'SubplotSpec' object has no attribute 'rowspan'
```

### PR DESCRIPTION
**Pull request by wakamex**
fixes #188
remove 3.11 tests since we're sticking with 3.8 for the foreseeable future for Colab compatibility

### CODE REVIEW

**Comment by wakamex** [comment]
along these lines, remove pin on numpy? @dpaiton

**Comment by wakamex** [comment]
progress

===== 3.8.6 =====
atomicwrites-1.4.1
attrs-22.2.0
cycler-0.11.0
kiwisolver-1.4.4
matplotlib-3.2.2
more-itertools-9.0.0
nodeenv-1.7.0
numpy-1.21.6
pandas-1.3.5
pluggy-0.7.1
py-1.11.0
pyparsing-3.0.9
pyright-1.1.293
pytest-3.6.4
python-dateutil-2.8.2
pytz-2022.7.1
scipy-1.7.3
six-1.16.0
stochastic-0.6.0

====== 3.11 ======
attrs-22.2.0
contourpy-1.0.7
cycler-0.11.0
fonttools-4.38.0
iniconfig-2.0.0
kiwisolver-1.4.4
matplotlib-3.6.3
nodeenv-1.7.0
packaging-23.0
pandas-1.5.3
pillow-9.4.0
pluggy-1.0.0
pyparsing-3.0.9
pyright-1.1.293
pytest-7.2.1
python-dateutil-2.8.2
pytz-2022.7.1
six-1.16.0
tomli-2.0.1

**Comment by wakamex** [comment]
the blazing fast speed
image

**Comment by wakamex** [comment]
python 3.11 was chugging hard trying to build obsolete versions of packages from 2 years ago from scratch
image

File: src/elfpy/utils/outputs.py [hunk 1]
```diff
@@ -327,7 +327,7 @@ def get_gridspec_subplots(nrows: int = 1, ncols: int = 1, **kwargs: Any) -> tupl
         kwargs["wspace"] = 1.0
     grid_spec = gridspec.GridSpec(nrows, ncols, **kwargs)
     fig = plt.figure()
-    axes = [fig.add_subplot(grid_spec[plot_id]) for plot_id in np.ndindex((nrows, ncols))]
+    axes = [fig.add_subplot(grid_spec[plot_id]) for plot_id in np.ndindex(nrows, ncols)]
```
Comment (wakamex) [inline src/elfpy/utils/outputs.py:330]: pyright complains only on 3.8.6 about this

File: requirements-3.8.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+numpy==1.21.6
+matplotlib==3.2.2
+pytest==3.6.4
+pylint>=2.15.10
```
Comment (dpaiton) [inline requirements-3.8.txt:4]: I _think_ it makes sense to put the less constrained (>=) requirements after the more constrained (==) requirements.

File: src/elfpy/utils/outputs.py (hunk 1 above)
Comment (dpaiton) [reply src/elfpy/utils/outputs.py:330]: goot catch, better anyway

**Review by dpaiton — APPROVED** [review]
LGTM -- make sure hyperdrive colab runs before merging!

**Pull request merged**